### ISSUE TITLE
Towards a configurable underlying floating point type

### ISSUE THREAD

**Issue by msmk0** [acts-project/acts#156]
At the moment, all code uses `double` as the underlying scalar floating point type. In the future, this should be changeable e.g. for performance reasons or to perform numerical tests.

As a first step, type-independent scalar and vector types should be introduced that can be used in new code instead of `double`, `Vector3D`, `Rotation3D`, ... .

```cpp
using ActsScalar = double; // can be made configurable later on
using ActsVector3 = ActsVector<Scalar, 3>;
using ActsVector4 = ActsVector<Scalar, 4>;
...
```

**Comment by paulgessinger** [comment]
I thought we discussed and decided against this at some point?

**Comment by HadrienG2** [comment]
Last time I did this work, the main problem was that no one was able to tell me what is "precise enough" for unit test tolerances. So I could switch to float, but couldn't tell if the result is good enough.

**Comment by msmk0** [comment]
> I thought we discussed and decided against this at some point?

I do not remember that we did, but I am not 100% sure about it. I was (re-)introducing this in the light of recent work on GPU prototypes. There we might need to use a different floating point type.

The idea here is to introduce basic support for such a change and then gradually move over parts of the code to be based on the type-independent tools. The specific change here is small and could be introduced right now, the larger change is a longer-term project.

**Comment by HadrienG2** [comment]
In case we do end up doing this, here's a little note regarding the practical side of things: from the experience of my past attempt, this kind of development branch bitrots **extremely fast** as it basically touches every function signature in the Acts codebase, which means that any incoming PR will cause merge conflicts.

So for a multi-precision port to be a success, you want to...

- Make sure everyone is onboard with the idea before doing the port.
- Save up some dedicated time in your week to perform the port as quickly as possible.
- Negotiate priorizing the port in front of every other outstanding PR in the review process.

Basically the same kind of things that we do when moving to a new clang-format release, as automated code formatting faces similar problems.
